### PR TITLE
RHDEVDOCS-4073 - Document on binding workloads using the label selector feature as per the spec 1.0

### PR DESCRIPTION
**Pull request by Srivaralakshmi**
[RHDEVDOCS-4073](https://issues.redhat.com//browse/RHDEVDOCS-4073): Document on binding workloads using the label selector feature as per the spec 1.0

- **Aligned team**: Dev Tools
- **OCP version for cherry-picking**: enterprise-`4.10` and later
- **JIRA issues**: [RHDEVDOCS-4073](https://issues.redhat.com/browse/RHDEVDOCS-4073)
- **Preview pages**: [Download to see the preview in your browser](https://drive.google.com/file/d/1t3ZSCZXhyBdN80rgN46CbRAIXwmIY_A-/view?usp=sharing)
- **SME Review**: Completed by @dperaza4dustbit, @sadlerap
- **QE review**:  Completed by @pmacik 
- **Peer-review**: Completed by @rolfedh

### CODE REVIEW

**Comment by Srivaralakshmi** [comment]
@dperaza4dustbit, @sadlerap, and @pmacik: PTAL at the PR and provide feedback. Thanks!

**Review by sadlerap — APPROVED** [review]
Looks good to me.

**Comment by Srivaralakshmi** [comment]
@sadlerap: Thanks a lot!

@dperaza4dustbit and @pmacik: PTAL at the PR and provide feedback. Thanks!

**Review by dperaza4dustbit — APPROVED** [review]
lgtm

**Comment by dperaza4dustbit** [comment]
We should be good to merge this

File: modules/sbo-advanced-binding-options.adoc [hunk 1]
```diff
@@ -7,10 +7,15 @@
...
+* Binding workloads using label selectors: This option is available for both the `binding.operators.coreos.com` and `servicebinding.io` API groups.
+
+[id="changing-binding-names_{context}"]
 == Changing the binding names before projecting them into the workload
-You can specify the rules to change the binding names in the .spec.namingStrategy attribute of the `ServiceBinding` custom resource (CR). For example, consider a Spring PetClinic sample application that connects to the PostgreSQL database. In this case, the PostgreSQL database service exposes the host and port fields of the database to use for binding. The Spring PetClinic sample application can access this exposed binding data through the binding names.
+You can specify the rules to change the binding names in the .spec.namingStrategy attribute of the `ServiceBinding` CR. For example, consider a Spring PetClinic sample application that connects to the PostgreSQL database. In this case, the PostgreSQL database service exposes the host and port fields of the database to use for binding. The Spring PetClinic sample application can access this exposed binding data through the binding names.
```
[the remark this answers is not in the document]
Comment (rolfedh) [reply modules/sbo-advanced-binding-options.adoc:18]: Check with the SME whether these field names are allcaps or not. For example "the `HOST` and `PORT` fields"

File: modules/sbo-advanced-binding-options.adoc [hunk 2]
```diff
@@ -115,4 +121,66 @@ spec:
 <1> Name of the backing service resource.
 <2> Optional identifier.
 <3> Generated JSON name that is to be projected as the file content or environment value. The JSON name contains the attributes of the backing service custom resource.
-<4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.
\ No newline at end of file
+<4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.
```
Comment (rolfedh) [inline modules/sbo-advanced-binding-options.adoc]: The "that is to be projected" wording sounds strange. It contains both future tense and passive voice.

File: modules/sbo-advanced-binding-options.adoc [hunk 3]
```diff
@@ -115,4 +121,66 @@ spec:
...
+
+[id="binding-workloads-using-a-label-selector_{context}"]
+== Binding workloads using a label selector
+You can use a label selector to specify the workload that is being bound. If you declare a service binding using the label selectors to pick up workloads, the {servicebinding-title} periodically attempts to find and bind new workloads that match the given label selector.
```
Comment (rolfedh) [inline modules/sbo-advanced-binding-options.adoc:128]: How about:
```suggestion
You can use a label selector to specify the workload to bind. If you declare a service binding using the label selectors to pick up workloads, the {servicebinding-title} periodically attempts to find and bind new workloads that match the given label selector.
```
This suggestion incorporates the following guidelines:
- Avoid passive voice.

File: modules/sbo-advanced-binding-options.adoc [hunk 4]
```diff
@@ -115,4 +121,66 @@ spec:
...
+<1> Specifies the workload that is being bound.
+
+[IMPORTANT]
+====
+Currently, it is forbidden to attempt a binding with the following fields defined:
+
+* Both `name` and `labelSelector` in the `binding.operators.coreos.com/v1alpha1` API
+* Both `name` and `selector` in the `servicebinding.io` API (Spec API)
+
+Any attempt to do so will result in an error.
```
Comment (rolfedh) [inline modules/sbo-advanced-binding-options.adoc:185]: Rephrase to use the present tense.

File: modules/sbo-advanced-binding-options.adoc [hunk 5]
```diff
@@ -115,4 +121,66 @@ spec:
...
+    resource: deployments
+  services:
+    group: ""
+    version: v1
+    kind: Secret
+    name: super-secret-data
+----
+<1> Specifies the workload that is being bound.
+
+.Example of `ServiceBinding` CR in the `servicebinding.io` API:
```
Comment (rolfedh) [inline modules/sbo-advanced-binding-options.adoc:156]:
```suggestion
.Example `ServiceBinding` CR in the `servicebinding.io` API
```

File: modules/sbo-advanced-binding-options.adoc (hunk 4 above)
Comment (rolfedh) [inline modules/sbo-advanced-binding-options.adoc:185]:
```suggestion
* The `name` and `labelSelector` fields in the `binding.operators.coreos.com/v1alpha1` API
* The `name` and `selector` fields in the `servicebinding.io` API (Spec API)
```

Comment (rolfedh) [reply modules/sbo-advanced-binding-options.adoc:185]: If you keep line 185, please rephrase it to avoid using future tense.

**Review by rolfedh — COMMENTED** [review]
Hi, @Srivaralakshmi. Nice work! I left some comments and suggestions.

File: modules/sbo-advanced-binding-options.adoc [hunk 6]
```diff
@@ -115,4 +121,66 @@ spec:
...
-<4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.
\ No newline at end of file
+<4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.
+
+[id="binding-workloads-using-a-label-selector_{context}"]
+== Binding workloads using a label selector
+You can use a label selector to specify the workload that is being bound. If you declare a service binding using the label selectors to pick up workloads, the {servicebinding-title} periodically attempts to find and bind new workloads that match the given label selector.
+
+For example, you may want to bind a service to every `Deployment` in a namespace with the `environment: production` label set in the `application` field. By setting an appropriate label selector, you can enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.
```
[the remark this answers is not in the document]
Comment (rolfedh) [reply modules/sbo-advanced-binding-options.adoc:130]:
```suggestion
For example, to bind a service to every `Deployment` in a namespace, you set the `environment: production` label as the value of the `application` field. By setting an appropriate label selector, you enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.
```

[the remark this answers is not in the document]
Comment (rolfedh) [reply modules/sbo-advanced-binding-options.adoc:130]: Or...
```suggestion
For example, you can bind a service to every `Deployment` in a namespace by using the `environment: production` label as the value of the `application` field. Setting the appropriate label selector enables the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.
```

File: modules/sbo-advanced-binding-options.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:124]: This is the version that I prefer and to the point:

"The JSON value that the {servicebinding-title} generates. The {servicebinding-title} projects this JSON value as a file or environment variable. The JSON value contains the attributes from the specified field of the backing service custom resource."

[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:124]: "The JSON name that the {servicebinding-title} generates. The {servicebinding-title} projects this JSON name as a file or environment variable. The JSON name contains the attributes from your specified field of the backing service custom resource."

File: modules/sbo-advanced-binding-options.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:130]: For example, as a cluster administrator, you can bind a service to every `Deployment` in a namespace with the `environment: production` label by setting an appropriate `labelSelector` field in the `ServiceBinding` CR.  This enables the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` CR.

File: modules/sbo-advanced-binding-options.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:18]: Fixed

File: modules/sbo-advanced-binding-options.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:130]: Fixed

File: modules/sbo-advanced-binding-options.adoc [hunk 7]
```diff
@@ -115,4 +121,66 @@ spec:
...
+
+.Example of `ServiceBinding` CR in the `binding.operators.coreos.com/v1alpha1` API:
```
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:132]: Fixed

File: modules/sbo-advanced-binding-options.adoc (hunk 5 above)
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:156]: Fixed

File: modules/sbo-advanced-binding-options.adoc [hunk 8]
```diff
@@ -115,4 +121,66 @@ spec:
...
+  service:
+    apiVersion: v1
+    kind: Secret
+    name: super-secret-data
+----
+<1> Specifies the workload that is being bound.
+
+[IMPORTANT]
+====
+Currently, it is forbidden to attempt a binding with the following fields defined:
```
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:180]: Fixed

File: modules/sbo-advanced-binding-options.adoc (hunk 4 above)
Comment (Srivaralakshmi) [reply modules/sbo-advanced-binding-options.adoc:185]: Fixed

**Comment by Srivaralakshmi** [comment]
@rolfedh Thanks a lot for the useful suggestions. I have incorporated all of them. PTAL.

**Review by rolfedh — APPROVED** [review]
/lgtm
